### PR TITLE
Swap functionality on Android

### PR DESCRIPTION
**Pull request by SergeyZhukovsky**
Resolves https://github.com/brave/brave-browser/issues/18336

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/common/brave_wallet.mojom [hunk 1]
```diff
@@ -248,7 +248,14 @@ interface AssetRatioController {
...
 };
 
+struct SwapConfig {
+  string swap_api_url;
+  string buy_token_percantage_fee;
+  string fee_recipient;
+};
+
 interface SwapController {
+  GetSwapConfiguration(string chain_id) => (SwapConfig config);
```
Comment (SergeyZhukovsky) [inline components/brave_wallet/common/brave_wallet.mojom:258]: Created a desktop issue to use that function https://github.com/brave/brave-browser/issues/18546

File: android/java/org/chromium/chrome/browser/crypto_wallet/SwapControllerFactory.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import org.chromium.mojo.bindings.ConnectionErrorHandler;
+import org.chromium.mojo.bindings.Interface;
+import org.chromium.mojo.bindings.Interface.Proxy.Handler;
+import org.chromium.mojo.system.MessagePipeHandle;
+import org.chromium.mojo.system.MojoException;
+import org.chromium.mojo.system.impl.CoreImpl;
+
+@JNINamespace("chrome::android")
+public class SwapControllerFactory {
+    private static final Object lock = new Object();
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/crypto_wallet/SwapControllerFactory.java:21]: static fields should be named starting with `s`

File: android/java/org/chromium/chrome/browser/crypto_wallet/SwapControllerFactory.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.chromium.chrome.browser.crypto_wallet;
+
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/crypto_wallet/SwapControllerFactory.java:7]: seems unused:
```
import org.chromium.base.annotations.CalledByNative
import org.chromium.mojo.system.MojoException
```

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java [hunk 4]
```diff
@@ -268,29 +485,27 @@ private int getIndexOf(Spinner spinner, String chainId) {
     private void adjustControls() {
         EditText toValueText = findViewById(R.id.to_value_text);
         TextView marketPriceValueText = findViewById(R.id.market_price_value_text);
-        TextView slippingToleranceValueText = findViewById(R.id.slipping_tolerance_value_text);
-        RadioGroup radioBuySendSwap = findViewById(R.id.buy_send_swap_type_radio_group);
+        // RadioGroup radioBuySendSwap = findViewById(R.id.buy_send_swap_type_radio_group);
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java:488]: is this commented variable left by purpose?

**Review by AlexeyBarabash — APPROVED** [review]
looks good except few mentioned minor styling points

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java (hunk 4 above)
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java:488]: Yes, it's for market/limit orders, we do it without that for now, but could be needed in the future

Comment (AlexeyBarabash) [reply android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java:488]: thanks :+1:

**Comment by SergeyZhukovsky** [comment]
All platforms CI passed, adding skipping everything but Android as further changes are only Android related.

File: browser/brave_wallet/BUILD.gn [hunk 5]
```diff
@@ -38,6 +38,7 @@ source_set("brave_wallet") {
     "//services/network/public/cpp",
     "//third_party/abseil-cpp:absl",
   ]
+
```
Comment (bbondy) [inline browser/brave_wallet/BUILD.gn:41]: nit: Better to interactive rebase out unnecessary changes in PR

File: components/brave_wallet/browser/swap_controller.cc [hunk 6]
```diff
@@ -117,6 +117,24 @@ GURL SwapController::GetTransactionPayloadURL(
...
+    swap_config->swap_api_url = kRopstenSwapAPIURL;
+    swap_config->buy_token_percantage_fee = kRopstenBuyTokenPercentageFee;
+    swap_config->fee_recipient = kRopstenFeeRecipient;
+  } else {
+    swap_config->swap_api_url = kSwapAPIURL;
+    swap_config->buy_token_percantage_fee = kBuyTokenPercentageFee;
+    swap_config->fee_recipient = kFeeRecipient;
+  }
+
+  std::move(callback).Run(std::move(swap_config));
```
Comment (bbondy) [inline components/brave_wallet/browser/swap_controller.cc:135]: It would be easy for here to add a unittest to test both of these cases with expected config here `components/brave_wallet/browser/swap_controller_unittest.cc`.

File: components/brave_wallet/common/brave_wallet.mojom (hunk 1 above)
Comment (bbondy) [reply components/brave_wallet/common/brave_wallet.mojom:258]: Maybe a return result of `SwapConfig?` here so absl::optional<mojom::SwapConfig>` so that you can return an optional result when there is no known swap config for the passed in chain.

File: components/brave_wallet/browser/swap_controller.cc [hunk 7]
```diff
@@ -117,6 +117,24 @@ GURL SwapController::GetTransactionPayloadURL(
...
+void SwapController::GetSwapConfiguration(
+    const std::string& chain_id,
+    GetSwapConfigurationCallback callback) {
+  auto swap_config = mojom::SwapConfig::New();
+
+  if (chain_id == mojom::kRopstenChainId) {
+    swap_config->swap_api_url = kRopstenSwapAPIURL;
+    swap_config->buy_token_percantage_fee = kRopstenBuyTokenPercentageFee;
+    swap_config->fee_recipient = kRopstenFeeRecipient;
+  } else {
```
Comment (bbondy) [inline components/brave_wallet/browser/swap_controller.cc:129]: I think better to differentiate between a known and unknown swap config for a given chain id.

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/Utils.java [hunk 8]
```diff
@@ -270,6 +281,41 @@ public static double fromHexWeiToGWEI(String number) {
...
+            BigInteger res = bigNumber.multiply(new BigInteger(multiplier));
+
+            return res.toString();
+        } catch (NumberFormatException ex) {
+        }
+
+        return "0";
+    }
+
+    public static double fromWei(String number) {
```
Comment (bbondy) [inline android/java/org/chromium/chrome/browser/crypto_wallet/util/Utils.java:306]: Can we have some unit tests for these utils?

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java [hunk 9]
```diff
@@ -158,7 +159,8 @@ private void setupView(View view) {
                         if (!success || values.length == 0) {
                             return;
                         }
-                        double value = Utils.fromHexWei(mTxInfo.txData.baseData.gasPrice);
+                        double value =
+                                Utils.fromHexWei(mTxInfo.txData.baseData.gasPrice) * gasLimit;
```
Comment (bbondy) [inline android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java:163]: Shouldn't these calculations be done in BigInteger and then converted only for display?

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java [hunk 10]
```diff
@@ -56,6 +65,10 @@
 
 public class BuySendSwapActivity extends AsyncInitializationActivity
         implements ConnectionErrorHandler, AdapterView.OnItemSelectedListener {
+    private final static String TAG = "BuySendSwapActivity";
+    private final static String ETHEREUM_CONTRACT_FOR_SWAP =
+            "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee";
```
Comment (bbondy) [inline android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java:70]: Why is this an invalid address needed? I'm guessing we do the same on desktop so just asking for me.

**Review by bbondy — CHANGES_REQUESTED** [review]
@onyb should probably also review the swap logic

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java (hunk 10 above)
Comment (onyb) [reply android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java:70]: 0x recommends using contract addresses for identifying assets, instead of symbols. I think it's because two assets can share the same ticker, but never the same contract address. `0xee...` is just how 0x identifies ETH.

File: components/brave_wallet/browser/swap_controller.cc (hunk 7 above)
Comment (SergeyZhukovsky) [reply components/brave_wallet/browser/swap_controller.cc:129]: It's just a copy from desktop https://github.com/brave/brave-core/pull/10215/files#diff-6dcef5d72bc94ad3e02ebfb7fb11418fb009cb9cb08dc34c52268d30066e1f6dR21. @onyb should I just return an empty object for other networks?

Comment (onyb) [reply components/brave_wallet/browser/swap_controller.cc:129]: Swap is disabled on all networks other than mainnet or ropsten. So the `else` case would never be used.

**Review by bbondy — APPROVED** [review]

**Comment by SergeyZhukovsky** [comment]
macOS CI failed on tests BraveSearchTestEnabled.DefaultAPIVisibleKnownHost and SearchEngineProviderP3ATest.DefaultSearchEngineP3A which is a known problem
